radixdlt/src/main/java/com/radixdlt/mempool/SubmissionControlImpl.java

```diff
@@ -28,55 +31,70 @@
 import org.json.JSONObject;
 import org.radix.atoms.events.AtomExceptionEvent;
 import org.radix.events.Events;
-import org.radix.validation.ConstraintMachineValidationException;
 
-import com.radixdlt.identifiers.AID;
 import com.radixdlt.atommodel.Atom;
-import com.radixdlt.constraintmachine.CMError;
 import com.radixdlt.engine.RadixEngine;
 import com.radixdlt.serialization.Serialization;
 
 class SubmissionControlImpl implements SubmissionControl {
 	private static final Logger log = LogManager.getLogger("submission");
 
 	private final Mempool mempool;
-	private final RadixEngine radixEngine;
+	private final RadixEngine<LedgerAtom> radixEngine;
 	private final Serialization serialization;
 	private final Events events;
+	private final AtomToLedgerAtomConverter converter;
 
 	@Inject
-	SubmissionControlImpl(Mempool mempool, RadixEngine radixEngine, Serialization serialization, Events events) {
+	SubmissionControlImpl(
+		Mempool mempool,
+		RadixEngine<LedgerAtom> radixEngine,
+		Serialization serialization,
+		Events events,
+		AtomToLedgerAtomConverter converter
+	) {
 		this.mempool = Objects.requireNonNull(mempool);
 		this.radixEngine = Objects.requireNonNull(radixEngine);
 		this.serialization = Objects.requireNonNull(serialization);
 		this.events = Objects.requireNonNull(events);
+		this.converter = Objects.requireNonNull(converter);
 	}
 
 	@Override
-	public void submitAtom(Atom atom) throws MempoolFullException, MempoolDuplicateException {
-		Optional<CMError> validationError = this.radixEngine.staticCheck(atom);
-		if (validationError.isPresent()) {
-			CMError error = validationError.get();
-			ConstraintMachineValidationException ex = new ConstraintMachineValidationException(atom, error.getErrMsg(), error.getDataPointer());
+	public void submitAtom(LedgerAtom atom) throws MempoolFullException, MempoolDuplicateException {
+		try {
+			this.radixEngine.staticCheck(atom);
+		} catch (RadixEngineException e) {
 			log.info(
-				"Rejecting atom {} with constraint machine error '{}' at '{}'.",
+				"Rejecting atom {} with error '{}' at '{}'.",
 				atom.getAID(),
-				error.getErrorDescription(),
-				error.getDataPointer()
+				e.getErrorCode(),
+				e.getDataPointer()
 			);
-			this.events.broadcast(new AtomExceptionEvent(ex, atom.getAID()));
-		} else {
-			this.mempool.addAtom(atom);
+			this.events.broadcast(new AtomExceptionEvent(e, atom.getAID()));
+			return;
 		}
+
+		this.mempool.addAtom(atom);
```

Comment:
I think this would be better inside the try block.  That way we would avoid the awkward flow in this method.